### ISSUE TITLE
Schema attributes with inheritance are dropped

### ISSUE THREAD

**Issue by swaechter** [micronaut-projects/micronaut-openapi#483]
### Steps to Reproduce

1.) Clone the example https://github.com/swaechter/openapiproblems
2.) Switch the version number in gradle.properties to 2.4.0 and run the application
3.) Check out the schema model `Owner` that has an attribute `vehicle` of the type Car/Bike
4.) Stop the application and update the version number 2.4.1. Run the application
5.) The schema attribute `vehicle` is now missing, but the schemas Car/Bike still exist. In addition, the Vehicle schema references Car/Bike via oneOf

### Expected Behaviour

The `vehicle` attribute shows up in the editor & is always referenced:

![MN3](https://user-images.githubusercontent.com/5626755/112749355-1f826a00-8fc2-11eb-8f77-a7d0305279ed.png)

![MN4](https://user-images.githubusercontent.com/5626755/112749358-21e4c400-8fc2-11eb-8ed7-c51d8dc818df.png)

### Actual Behaviour

Starting with 2.4.1, the `vehicle` attribute doesn't show up in the editor & isn't referenced anymore. I am not sure if this is somehow related to @JsonXXX attributes that aren't processed properly (Looking at https://github.com/micronaut-projects/micronaut-openapi/issues/471)

![MN1](https://user-images.githubusercontent.com/5626755/112749304-d6cab100-8fc1-11eb-8b7c-d4e1559dcb9d.PNG)

![MN2](https://user-images.githubusercontent.com/5626755/112749310-dcc09200-8fc1-11eb-855b-125212241d5d.png)

### Environment Information

- **Operating System**: Windows 10 x64
- **Micronaut Version**: Working in 2.4.0, broken in 2.4.1+
- **JDK Version**: AdoptJDK jdk-11.0.10.9-hotspot

### Example Application

https://github.com/swaechter/openapiproblems

**Comment by swaechter** [comment]
It looks like the schema generation itself is correct. What's broken are the schema references (Schema X has an attribute Y of type Z --> Relationship Z has always the type/value "string" or "object" without $ref). I found these two broken relationships:

* anyOf: An Owner can jave a Car OR Bike, but it has to be of type Vehicle/contain all values of type Vehicle)
* $ref for nested classes or enums: ReadBackupSettingsDto has a backup frequency (Enum with weekday) and a time (Nested class with hour/minutes)

This means the generated OpenAPI code is unusable/broken as soon someone is using nested classes/enums :(

A workaround via custom merged YAML file is quite time consuming (Our YML is around 11k lines long) + using Micronaut 2.4.2 with an older annotation processor/library like 2.2.0 didn't work and 2.1.1 generates some strange BOM mistake (Unable to find ktor-bom on Maven central).

I hope I can find some time to dig into the problem (I always create issues without a PR). Do you know @asodja if the problem is related to your changes and maybe give a tip where to look?

![Diff](https://user-images.githubusercontent.com/5626755/116133149-ab290c80-a6ce-11eb-833a-110332473fbf.png)

**Comment by asodja** [comment]
It should not be related, since components ordering is done just before yaml is written to file, so my change does not affect actual schema referencing/creation. This change to you example also proves it:
```
diff --git a/build.gradle b/build.gradle
index 9e86e3d..0b11b62 100644
--- a/build.gradle
+++ b/build.gradle
@@ -22,8 +22,8 @@ micronaut {
 
 dependencies {
     annotationProcessor("io.micronaut:micronaut-management")
-    annotationProcessor("io.micronaut.openapi:micronaut-openapi")
-    implementation("io.micronaut.openapi:micronaut-openapi")
+    annotationProcessor("io.micronaut.openapi:micronaut-openapi:2.4.0!!")
+    implementation("io.micronaut.openapi:micronaut-openapi:2.4.0!!")
     implementation("io.swagger.core.v3:swagger-annotations")
     implementation("io.micronaut:micronaut-http-client")
     implementation("io.micronaut:micronaut-runtime")
diff --git a/gradle.properties b/gradle.properties
index be22b84..06f1798 100644
--- a/gradle.properties
+++ b/gradle.properties
@@ -1 +1 @@
-micronautVersion=2.4.2
+micronautVersion=2.4.0

```

So if you force latest Micronaut openapi that was released yesterday with my change and change Micronaut version to 2.4.0 everything works ok. If you change Micronaut version to >= 2.4.1 schema is broken again. 

So it looks like something in Micronaut core was changed and affects OpenApi.

----

I did not look into schema code much since I did not fix any issue regarding this. But if I remember correctly you should check AbstractOpenApiEndpointVisitor.java and method: https://github.com/micronaut-projects/micronaut-openapi/blob/9d9f092b0ea6f34d1beea8102d757ec58c8e9143/openapi/src/main/java/io/micronaut/openapi/visitor/AbstractOpenApiEndpointVisitor.java#L254 that later calls `processParameters()` or some similar method. Also check places where `Schema` class is used. 

Most of the code is in this class. Probably the easiest way to debug this is to set breakpoint in that method and then run in `OpenApiSchemaInheritanceSpec` debug mode and check what is happening here or implement your case and debug it.

I hope that helps.

**Comment by asodja** [comment]
Also it might be worth debugging, what is going on in:
https://github.com/micronaut-projects/micronaut-openapi/blob/9d9f092b0ea6f34d1beea8102d757ec58c8e9143/openapi/src/main/java/io/micronaut/openapi/visitor/AbstractOpenApiVisitor.java#L611

**Comment by graemerocher** [comment]
It is quite easy to add tests for this project. See https://github.com/micronaut-projects/micronaut-openapi/blob/master/openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiComplexSchemaSpec.groovy for an example if you want to come up with a reproducer

**Comment by asodja** [comment]
I had some time and played with this a bit, and the issue is this line where annotation on property and existing schema are merged:
https://github.com/micronaut-projects/micronaut-openapi/blob/9d9f092b0ea6f34d1beea8102d757ec58c8e9143/openapi/src/main/java/io/micronaut/openapi/visitor/AbstractOpenApiVisitor.java#L938 

It looks like it behaves differently with new Micronaut, it might be due to Jackson upgrade.

This will fix your issue:
```
diff --git a/src/main/java/ch/swaechter/dto/Owner.java b/src/main/java/ch/swaechter/dto/Owner.java
index c2df9bc..ab82b37 100644
--- a/src/main/java/ch/swaechter/dto/Owner.java
+++ b/src/main/java/ch/swaechter/dto/Owner.java
@@ -36,7 +36,7 @@ public class Owner {
         this.id = id;
     }
 
-    @Schema(description = "Vehicle of the owner. Here a car or bike with a name", oneOf = {Car.class, Bike.class})
+//    @Schema(description = "Vehicle of the owner. Here a car or bike with a name", oneOf = {Car.class, Bike.class})
     public Vehicle getVehicle() {
         return vehicle;
     }
diff --git a/src/main/java/ch/swaechter/dto/Vehicle.java b/src/main/java/ch/swaechter/dto/Vehicle.java
index ac98b96..8161412 100644
--- a/src/main/java/ch/swaechter/dto/Vehicle.java
+++ b/src/main/java/ch/swaechter/dto/Vehicle.java
@@ -14,6 +14,6 @@ import io.swagger.v3.oas.annotations.media.Schema;
         @JsonSubTypes.Type(value = Bike.class, name = "BIKE")
 })
 @Introspected
-@Schema(description = "Vehicle, here a car or a vehicle")
+@Schema(description = "Vehicle, here a car or a vehicle", anyOf = { Car.class, Bike.class })
 public abstract class Vehicle {
 }
```

So basically: for classes you have to have `@Schema` on class and not on getter/field to make it work.

I probably won't have time to fix it, but here is a reproducer if you want to find a solution.
https://github.com/asodja/micronaut-openapi/commit/f0491c78ffd62152798e30c1b129fe9be96c1bbe
